core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,433 @@
+                    upstreamRequested--;
+                }
+            }
+
+            if (!publishedAny) {
+                publishedAny = true;
+            }
+
+            publish();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (state == State.DONE) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            publishDownstream(item, false);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            close(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            setState(State.REQUIRE_BODY, State.REQUIRE_TAIL);
+            if (tail != null) {
+                publish();
+            } else {
+                close0(null);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (state == State.DONE) {
+                return;
+            }
+            state = State.DONE;
+
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+                upstream.cancel();
+            }
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (containsNotifyCancellation(options)) {
+                downstream.onError(cause);
+            }
+            downstream = NoopSubscriber.get();
+            completionFuture.completeExceptionally(cause);
+            release();
+        }
+
+        private void close(@Nullable Throwable cause) {
+            if (executor.inEventLoop()) {
+                close0(cause);
+            } else {
+                executor.execute(() -> close0(cause));
+            }
+        }
+
+        private void close0(@Nullable Throwable cause) {
+            if (state == State.DONE) {
+                return;
+            }
+            state = State.DONE;
+
+            if (cause == null) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                final Subscription upstream = this.upstream;
+                if (upstream != null) {
+                    upstream.cancel();
+                }
+                downstream.onError(cause);
+                completionFuture.completeExceptionally(cause);
+            }
+            release();
+        }
+
+        private void release() {
+            if (head != null) {
+                PooledObjects.close(head);
```

Comment:
Should we use `StreamMessageUtil.closeOrAbort()` instead which will handle more well-known types? If this class is a subclass of `HttpMessage`, `PooledObjects.close()` would be enough. However, this is an implementation of `StreaMessage` that can contain arbitrary object types.
```suggestion
                StreamMessageUtil.close(head, cause);
```

Reply:
✅ updated